.github/workflows/ci.yml

```diff
@@ -5,24 +5,98 @@ jobs:
+        echo "BASE_IMAGE_TAGS=$BASE_IMAGE_NAME:$TAG_NAME$maybe_base_image_latest_tag" >> $GITHUB_ENV
+        maybe_builder_image_latest_tag=${{ fromJSON(env.ON_DEFAULT_BRANCH) && ',$BUILDER_IMAGE_NAME:latest' || '' }}
+        echo "BUILDER_IMAGE_TAGS=$BUILDER_IMAGE_NAME:$TAG_NAME$maybe_builder_image_latest_tag" >> $GITHUB_ENV
+
+    - if: ${{ fromJSON(env.TAG_AND_PUSH) }}
       uses: docker/login-action@v1
       with:
         username: ${{ secrets.DOCKER_LOGIN }}
         password: ${{ secrets.DOCKER_PASSWORD }}
-    - if: ${{ github.event_name != 'pull_request' && github.ref_type == 'branch' && github.ref_name == github.event.repository.default_branch }}
-      run: |
-        ./devel/tag build-$CACHE_DATE
-        ./devel/push latest build-$CACHE_DATE
-    - if: ${{ github.event_name != 'pull_request' && github.ref_type == 'branch' && github.ref_name != github.event.repository.default_branch }}
-      # From `man docker-image-tag`: The tag name must be valid ASCII and may
-      # contain lowercase and uppercase letters, digits, underscores, periods
-      # and hyphens.
-      run: |
-        tag=branch-${GITHUB_REF_NAME//[^A-Za-z0-9._-]/-}
-        ./devel/tag $tag
-        ./devel/push $tag
+
+    # The nextstrain/base Dockerfile is a multi-stage with both a "builder" target
+    # and a main target.  To enable proper caching via --cache-from we need both
+    # these images available to pull layers from.  This means pulling both in at
+    # the start and pushing both up at the end.
+
+    # Calling `docker run nextstrain/base` will still only pull down the small base
+    # image rather than pulling down the larger nextstrain/base-builder image.
+    - name: Build${{ fromJSON(env.TAG_AND_PUSH) && '/tag/push' || '' }} ${{ env.BUILDER_IMAGE_NAME }}
+      uses: docker/build-push-action@v2
+      with:
+        target: builder
+        builder: ${{ steps.builder-image-builder.outputs.name }}
+        platforms: linux/amd64
+        context: .
+        pull: true
+        push: ${{ fromJSON(env.TAG_AND_PUSH) }}
+        tags: ${{ fromJSON(env.TAG_AND_PUSH) && env.BUILDER_IMAGE_TAGS || null }}
+        cache-from: |
+          type=registry,ref=${{ env.BUILDER_IMAGE_NAME }}
+          type=registry,ref=${{ env.BASE_IMAGE_NAME }}
+        build-args: |
+          CACHE_DATE
+          GIT_REVISION
+
+    - name: Build${{ fromJSON(env.TAG_AND_PUSH) && '/tag/push' || '' }} ${{ env.BASE_IMAGE_NAME }}
+      uses: docker/build-push-action@v2
+      with:
+        builder: ${{ steps.base-image-builder.outputs.name }}
+        platforms: linux/amd64
+        context: .
+        pull: true
+        push: ${{ fromJSON(env.TAG_AND_PUSH) }}
+        tags: ${{ fromJSON(env.TAG_AND_PUSH) && env.BASE_IMAGE_TAGS || null }}
+        cache-from: |
+          type=registry,ref=${{ env.BUILDER_IMAGE_NAME }}
+          type=registry,ref=${{ env.BASE_IMAGE_NAME }}
```

Comment:
Similarly, the lack of non-registry sources in this `cache-from` here might mean that this step can't use the just-built first stage from the previous step when `TAG_AND_PUSH` is false (and would have to re-download it with `TAG_AND_PUSH` is true).